cis194/week2/robert-chang/src/LogAnalysis.hs

```diff
@@ -53,3 +53,29 @@ parse :: String -> [LogMessage]
 parse [] = []
 parse file = map parseMessage $ lines file
 
+
+
+-- Exercise 2
+
+insert :: LogMessage -> MessageTree -> MessageTree
+{- if insert is given a LogMessage which is
+Unknown, it should return the MessageTree unchanged
+-}
+insert (Unknown _) tree = tree 
+{- insert to empty tree, should have LogMessage inserted -} 
+insert x Leaf        = Node Leaf x Leaf 
+insert x@(LogMessage _ t1 _) (Node left xs@(LogMessage _ t2 _) right) 
+  | t1 == t2 = Node left xs right
+  | t1 < t2  = Node (insert x left) xs right
+  | t1 > t2  = Node left xs (insert x right)
```

Comment:
Do you have a warning from GHC about non-exhaustive pattern match? If you do, please think about of a way to fix it because it means you have a partial function.

Reply:
It's a common practice to use `-Wall` but I think you can pass it when you compile/type checking instead of on the file itself. The same for  `NoImplicitPrelude`. They are both project wide options hence setting them on the complication/type checking step makes more sense.

Reply:
what about this:
```js
insert (LogMessage _ _ _) (Node _ (Unknown _) _) = undefined
insert (LogMessage _ _ _) (Node _ (LogMessage _ _ _) _) = undefined
```

Reply:
To keep a record here that using undefined makes this function partial.